File: graphics/examplesPython/bluntStr.py
#!/usr/bin/env python
import os
try:
  VTK_DATA = os.environ['VTK_DATA']
except KeyError:
  VTK_DATA = '../../../vtkdata/'

from libVTKCommonPython import *
from libVTKGraphicsPython import *

# Create dashed streamlines

from colors import *
# Create the RenderWindow, Renderer and both Actors
#
ren = vtkRenderer()
renWin = vtkRenderWindow()
renWin.AddRenderer(ren)
iren = vtkRenderWindowInteractor()
iren.SetRenderWindow(renWin)

# read data
#
pl3d = vtkPLOT3DReader()
pl3d.SetXYZFileName(VTK_DATA + "/bluntfinxyz.bin")
pl3d.SetQFileName(VTK_DATA + "/bluntfinq.bin")
pl3d.SetScalarFunctionNumber(100)
pl3d.SetVectorFunctionNumber(202)
pl3d.Update()

#streamers
#
seeds = vtkLineSource()
seeds.SetResolution(25)
seeds.SetPoint1(-6.5,0.25,0.10)
seeds.SetPoint2(-6.5,0.25,5.0)
streamers = vtkDashedStreamLine()
streamers.SetInput(pl3d.GetOutput())
streamers.SetSource(seeds.GetOutput())
streamers.SetMaximumPropagationTime(25)
streamers.SetStepLength(0.25)
streamers.Update()
mapStreamers = vtkPolyDataMapper()
mapStreamers.SetInput(streamers.GetOutput())
mapStreamers.SetScalarRange(  \
pl3d.GetOutput().GetPointData().GetScalars().GetRange() )
streamersActor = vtkActor()
streamersActor.SetMapper(mapStreamers)

# wall
#
wall = vtkStructuredGridGeometryFilter()
wall.SetInput(pl3d.GetOutput())
wall.SetExtent(0,100,0,0,0,100)
wallMap = vtkPolyDataMapper()
wallMap.SetInput(wall.GetOutput())
wallMap.ScalarVisibilityOff()
wallActor = vtkActor()
wallActor.SetMapper(wallMap)
wallActor.GetProperty().SetColor(0.8,0.8,0.8)

# fin
# 
fin = vtkStructuredGridGeometryFilter()
fin.SetInput(pl3d.GetOutput())
fin.SetExtent(0,100,0,100,0,0)
finMap = vtkPolyDataMapper()
finMap.SetInput(fin.GetOutput())
finMap.ScalarVisibilityOff()
finActor = vtkActor()
finActor.SetMapper(finMap)
finActor.GetProperty().SetColor(0.8,0.8,0.8)

# outline
outline = vtkStructuredGridOutlineFilter()
outline.SetInput(pl3d.GetOutput())
outlineMapper = vtkPolyDataMapper()
outlineMapper.SetInput(outline.GetOutput())
outlineActor = vtkActor()
outlineActor.SetMapper(outlineMapper)
outlineProp=outlineActor.GetProperty()
outlineProp.SetColor(1,1,1)

# Add the actors to the renderer, set the background and size
#
ren.AddActor(outlineActor)
ren.AddActor(streamersActor)
ren.AddActor(wallActor)
ren.AddActor(finActor)
ren.SetBackground(0,0,0)
renWin.SetSize(700,500)

cam1 = vtkCamera()
cam1.SetFocalPoint(2.87956,4.24691,2.73135)
cam1.SetPosition(-3.46307,16.7005,29.7406)
cam1.SetViewAngle(30)
cam1.SetViewUp(0.127555,0.911749,-0.390441)
ren.SetActiveCamera(cam1)

iren.Initialize()
renWin.Render()

# render the image
#

renWin.Render()


iren.Start()

File: graphics/examplesCxx/bluntStr.cxx
#include "vtkRenderer.h"
#include "vtkRenderWindow.h"
#include "vtkRenderWindowInteractor.h"
#include "vtkPLOT3DReader.h"
#include "vtkStructuredGridOutlineFilter.h"
#include "vtkPolyDataMapper.h"
#include "vtkActor.h"
#include "vtkStructuredGridGeometryFilter.h"
#include "vtkLineSource.h"
#include "vtkDashedStreamLine.h"

#include "SaveImage.h"

int main( int argc, char *argv[] )
{
  float range[2], c[3];
  float maxVelocity, maxTime = 0;
  
  vtkRenderer *aren = vtkRenderer::New();
  vtkRenderWindow *renWin = vtkRenderWindow::New();
    renWin->AddRenderer(aren);
  vtkRenderWindowInteractor *iren = vtkRenderWindowInteractor::New();
    iren->SetRenderWindow(renWin);

  vtkPLOT3DReader *reader = vtkPLOT3DReader::New();
    //reader->DebugOn();
    reader->SetXYZFileName("../../../vtkdata/bluntfinxyz.bin");
    reader->SetQFileName("../../../vtkdata/bluntfinq.bin");
    reader->SetFileFormat(VTK_WHOLE_SINGLE_GRID_NO_IBLANKING);
    reader->Update(); //force a read to occur
    reader->GetOutput()->GetCenter(c);
    reader->GetOutput()->GetLength();
  if ( reader->GetOutput()->GetPointData()->GetScalars() )
    {
    reader->GetOutput()->GetPointData()->GetScalars()->GetRange(range);
    }
  if ( reader->GetOutput()->GetPointData()->GetVectors() )
    {
    maxVelocity = reader->GetOutput()->GetPointData()->GetVectors()->GetMaxNorm();
    maxTime = 6.0*reader->GetOutput()->GetLength()/maxVelocity ;
    }

  vtkStructuredGridOutlineFilter *outlineF = vtkStructuredGridOutlineFilter::New();
    outlineF->SetInput(reader->GetOutput());
  vtkPolyDataMapper *outlineMapper = vtkPolyDataMapper::New();
    outlineMapper->SetInput(outlineF->GetOutput());
  vtkActor *outline = vtkActor::New();
    outline->SetMapper(outlineMapper);
    outline->GetProperty()->SetColor(1.0,1.0,1.0);
//
// Some geometry for context
//
  vtkStructuredGridGeometryFilter *wall = vtkStructuredGridGeometryFilter::New();
    wall->SetInput(reader->GetOutput());
    wall->SetExtent(0, 100, 0, 100, 0, 0);
  vtkPolyDataMapper *wallMap = vtkPolyDataMapper::New();
    wallMap->SetInput(wall->GetOutput());
    wallMap->ScalarVisibilityOff();
  vtkActor *wallActor = vtkActor::New();
    wallActor->SetMapper(wallMap);
    wallActor->GetProperty()->SetColor(0.2,0.2,0.2);

  vtkStructuredGridGeometryFilter *fin = vtkStructuredGridGeometryFilter::New();
    fin->SetInput(reader->GetOutput());
    fin->SetExtent(0,100,0,0,0,100);
  vtkPolyDataMapper *finMap = vtkPolyDataMapper::New();
    finMap->SetInput(fin->GetOutput());
    finMap->ScalarVisibilityOff();
  vtkActor *finActor = vtkActor::New();
    finActor->SetMapper(finMap);
    finActor->GetProperty()->SetColor(0.4, 0.4, 0.4);
//
// regular streamlines
//
  vtkLineSource *line1 = vtkLineSource::New();
    line1->SetResolution(25);
    line1->SetPoint1(-6.36, 0.25, 0.06);
    line1->SetPoint2(-6.36, 0.25, 5.37);
  vtkPolyDataMapper *rakeMapper = vtkPolyDataMapper::New();
    rakeMapper->SetInput(line1->GetOutput());
  vtkActor *rake1 = vtkActor::New();
    rake1->SetMapper(rakeMapper);
    rake1->GetProperty()->SetColor(1.0,1.0,1.0);

  vtkDashedStreamLine *streamers = vtkDashedStreamLine::New();
  //  streamers->DebugOn();
    streamers->SetInput(reader->GetOutput());
    streamers->SetSource(line1->GetOutput());
    streamers->SetMaximumPropagationTime(maxTime);
    streamers->SetStepLength(maxTime/150.0);
    streamers->SetDashFactor(0.50);
    streamers->SetIntegrationStepLength(0.2);
    streamers->Update();
    //streamers->DebugOff();

  vtkPolyDataMapper *streamersMapper = vtkPolyDataMapper::New();
    streamersMapper->SetInput(streamers->GetOutput());
    streamersMapper->SetScalarRange(range);

  vtkActor *lines = vtkActor::New();
    lines->SetMapper(streamersMapper);

  aren->AddActor(outline);
  aren->AddActor(wallActor);
  aren->AddActor(finActor);
  aren->AddActor(rake1);
  aren->AddActor(lines);
  aren->SetBackground(0.0,0.0,0.0);
  aren->GetActiveCamera()->Elevation(30.0);
  aren->GetActiveCamera()->Azimuth(30.0);
  aren->GetActiveCamera()->Zoom(2.0);
  aren->GetActiveCamera()->SetClippingRange(1,1000);

  renWin->SetSize(300,150);
  renWin->Render();

  SAVEIMAGE( renWin );

  // interact with data
  iren->Start();

  // Clean up
  aren->Delete();
  renWin->Delete();
  iren->Delete();
  reader->Delete();
  outlineF->Delete();
  outlineMapper->Delete();
  outline->Delete();
  wall->Delete();
  wallMap->Delete();
  wallActor->Delete();
  fin->Delete();
  finMap->Delete();
  finActor->Delete();
  line1->Delete();
  rakeMapper->Delete();
  rake1->Delete();
  streamers->Delete();
  streamersMapper->Delete();
  lines->Delete();

  return( 0 );
}
